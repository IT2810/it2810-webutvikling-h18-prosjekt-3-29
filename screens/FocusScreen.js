import React from 'react';
import MapView from 'react-native-maps';
import { Platform, View, StyleSheet, AsyncStorage } from 'react-native';
import { Location, Permissions } from 'expo';
import { IntentLauncherAndroid } from 'expo';
import MapStyle from './mapStyle.json';

export default class FocusScreen extends React.Component {
  state = {
    focusPoints: "",
    region: {},
    validLat: 63.4156144, //A-blokka at Gløshaugen
    validLong: 10.4045326, //A-blokka at Gløshaugen
    errorMessage: null,
  };
  
  //Set Title on Page
  static navigationOptions = {
    title: 'Gløshaugen Focus',
  };

  //If Android device, ask for LocationService permissions, anyhow -> fetch current location
  componentWillMount() {
    if (Platform.OS === 'android') {
      this._checkProviderAsync();
    } 
    this._getLocationAsync();
  }

  //Check location every minute to make sure user is within parameters
  componentDidMount(){
    this._getFocusPointsAsync();
    this.interval = setInterval(() => {
      this._getLocationAsync();
    }, 60000);
  }

    /* For Android 6. we need to specifically override LocationServices at first launch 
       due to incompatibility between expo.cli and Android SDK
    */
  _checkProviderAsync = async () => {
    let {status } = await Expo.Location.getProviderStatusAsync();
    if(status.locationServicesEnabled === 'false' || status.gpsAvailable === 'false'){
        IntentLauncherAndroid.startActivityAsync(
          IntentLauncherAndroid.ACTION_LOCATION_SOURCE_SETTINGS
        ); 
    } 
  };

  /* Save points to Async Storage */
  _saveFocusPointsAsync = async (value) => {
    let points = String(this.state.focusPoints + 1)
      try {
        await AsyncStorage.setItem('focusPoints', points);
        this.setState({focusPoints : parseInt(points)})
      } catch (error) {
        console.log(error.message);
      }
    };

    /* Get current point score from Async Storage*/
  _getFocusPointsAsync = async () => {
    try {
      let focusPointsFetched = await AsyncStorage.getItem('focusPoints');
      if (focusPointsFetched == null) {
        this.setState({focusPoints : 0})
      }
      else {
        focusPointsFetched = parseInt(focusPointsFetched)
        this.setState({focusPoints : focusPointsFetched})
      }
    } catch (error) {
      console.log(error.message);
    }
  }

  /* Fetch current position using expo-api, set this position to state
  /* Calculate distance from current location to known valid location, 
  /* if distance is too large -> dont give points (dont save)
  */
  _getLocationAsync = async () => {
    let { status } = await Permissions.askAsync(Permissions.LOCATION);
    if(status === 'granted'){
      console.log("Permission granted")
    }
    if (status !== 'granted') {
      this.setState({
        errorMessage: 'Permission to access location was denied',
      });
    }

    let location = await Location.getCurrentPositionAsync({enableHighAccuracy: true});
    let currentLat = location.coords.latitude;
    let currentLong = location.coords.longitude;
    let distanceNow = getDistanceFromLatLonInKm(currentLat,
      currentLong,
      this.state.validLat,
      this.state.validLong)

    if(distanceNow >= 0.05){
      alert("You are currently not at Gløshuagen (A-blokka)! Get 1 point for every minute you spend there! Study hard, get smarter!")
    } else {
      alert("Yay, you are at Gløshaugen (A-blokka)! Get 1 point for every minute you spend there! Study hard, get smarter!")
      this._saveFocusPointsAsync();
    }

    this.setState({region: {
      latitude:       currentLat,
      longitude:      currentLong,
      latitudeDelta:  0.00922*1.6, //Zoom level
      longitudeDelta: 0.00421*1.6, //Zoom level
    }})  
  };

  render() {
    /* If one wants conditional styles on map based on OS, can be added later */
    let mapStylePlatform = "";
    if (Platform.OS === 'android') {
      mapStylePlatform = MapStyle;
    } else {
      mapStylePlatform = MapStyle;
    }
    return (
      <View style={styles.container}>
        <MapView
          provider={MapView.PROVIDER_GOOGLE}
          style={styles.map}
          customMapStyle={mapStylePlatform}
          region={this.state.region}
          showsUserLocation={true}
          followUserLocation={true}>
        </MapView>
      </View>
    );
  }
}

/* Common implementation of Haversine Formula */
function getDistanceFromLatLonInKm(lat1,lon1,lat2,lon2) {
  var R = 6371; // Radius of the earth in km
  var dLat = deg2rad(lat2-lat1);  // deg2rad below
  var dLon = deg2rad(lon2-lon1); 
  var a = 
    Math.sin(dLat/2) * Math.sin(dLat/2) +
    Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) * 
    Math.sin(dLon/2) * Math.sin(dLon/2)
    ; 
  var c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a)); 
  var d = R * c; // Distance in km
  return d;
}

function deg2rad(deg) {
  return deg * (Math.PI/180)
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    justifyContent: 'flex-end',
    alignItems: 'center'
  },
  map: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  }
});
