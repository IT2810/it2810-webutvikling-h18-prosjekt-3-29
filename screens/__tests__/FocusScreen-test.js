import React from 'react';
import FocusScreen from '../FocusScreen';
import renderer from 'react-test-renderer';
import MockAsyncStorage from 'mock-async-storage';
import { AsyncStorage as storage } from 'react-native'

//Implements mocks, the remianing mocks are done in test.setup.js
const mock = () => {
    const mockImpl = new MockAsyncStorage()
    jest.mock('AsyncStorage', () => mockImpl)
  }
  
  it('should render corrrectly', async () => {
    mock();
    const tree = renderer.create(<FocusScreen/>).toJSON();
    expect (tree).toMatchSnapshot();
    jest.unmock('AsyncStorage');
  })