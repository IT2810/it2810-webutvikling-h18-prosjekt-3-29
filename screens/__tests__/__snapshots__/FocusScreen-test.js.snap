// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render corrrectly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "justifyContent": "flex-end",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <MapView
    customMapStyle={
      Array [
        Object {
          "elementType": "all",
          "featureType": "all",
          "stylers": Array [
            Object {
              "visibility": "on",
            },
          ],
        },
        Object {
          "elementType": "labels",
          "featureType": "all",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
            Object {
              "saturation": "-100",
            },
          ],
        },
        Object {
          "elementType": "labels.text.fill",
          "featureType": "all",
          "stylers": Array [
            Object {
              "saturation": 36,
            },
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 40,
            },
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "labels.text.stroke",
          "featureType": "all",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 16,
            },
          ],
        },
        Object {
          "elementType": "labels.icon",
          "featureType": "all",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "administrative",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 20,
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "administrative",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 17,
            },
            Object {
              "weight": 1.2,
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "landscape",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 20,
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "landscape",
          "stylers": Array [
            Object {
              "color": "#4d6059",
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "landscape",
          "stylers": Array [
            Object {
              "color": "#4d6059",
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "landscape.natural",
          "stylers": Array [
            Object {
              "color": "#4d6059",
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "poi",
          "stylers": Array [
            Object {
              "lightness": 21,
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "poi",
          "stylers": Array [
            Object {
              "color": "#4d6059",
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "poi",
          "stylers": Array [
            Object {
              "color": "#4d6059",
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "road",
          "stylers": Array [
            Object {
              "visibility": "on",
            },
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "road",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "road.highway",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
            Object {
              "lightness": 17,
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "road.highway",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
            Object {
              "lightness": 29,
            },
            Object {
              "weight": 0.2,
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "road.arterial",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 18,
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "road.arterial",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "road.arterial",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "road.local",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 16,
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "road.local",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "road.local",
          "stylers": Array [
            Object {
              "color": "#7f8d89",
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "transit",
          "stylers": Array [
            Object {
              "color": "#000000",
            },
            Object {
              "lightness": 19,
            },
          ],
        },
        Object {
          "elementType": "all",
          "featureType": "water",
          "stylers": Array [
            Object {
              "color": "#2b3638",
            },
            Object {
              "visibility": "on",
            },
          ],
        },
        Object {
          "elementType": "geometry",
          "featureType": "water",
          "stylers": Array [
            Object {
              "color": "#2b3638",
            },
            Object {
              "lightness": 17,
            },
          ],
        },
        Object {
          "elementType": "geometry.fill",
          "featureType": "water",
          "stylers": Array [
            Object {
              "color": "#24282b",
            },
          ],
        },
        Object {
          "elementType": "geometry.stroke",
          "featureType": "water",
          "stylers": Array [
            Object {
              "color": "#24282b",
            },
          ],
        },
        Object {
          "elementType": "labels",
          "featureType": "water",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "labels.text",
          "featureType": "water",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "labels.text.fill",
          "featureType": "water",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "labels.text.stroke",
          "featureType": "water",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
        Object {
          "elementType": "labels.icon",
          "featureType": "water",
          "stylers": Array [
            Object {
              "visibility": "off",
            },
          ],
        },
      ]
    }
    followUserLocation={true}
    region={Object {}}
    showsUserLocation={true}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;
